tiltfile/dockerpruner: configure docker pruner from tiltfile, more options for when we prune [ch3661]

maiamcc opened the pull request:
Hello @jazzdan, @nicks,

Please review the following commits I made in branch maiamcc/tiltfile-docker-prune:

3b8ef4389c13c4b58289ce0bfa89b913961d6e18 (2019-10-11 17:17:45 -0400)
tests

d13689981265d54f895b1e696d422dd889048dc5 (2019-10-11 17:17:42 -0400)
be smarter about when we do/don't prune

6a375340c0e32f3edc5022cbde4b00df80e68b3f (2019-10-11 17:17:38 -0400)
tiltfile tweaks

dd885d0515e2a034026146425790f02eac640034 (2019-10-11 17:17:35 -0400)
apply the settings

94e8746c7b823579f7a8972de6169a5da231db76 (2019-10-11 17:17:30 -0400)
control docker prune settings from tiltfile

Code review reminders, by giving a LGTM you attest that:

* Commits are adequately tested
* Code is easy to understand and conforms to style guides
* Incomplete code is marked with TODOs
* Code is suitably instrumented with logging and metrics

nicks commented on internal/tiltfile/tiltfile_state.go:
```diff
@@ -1226,6 +1235,34 @@ func (s *tiltfileState) setTeam(thread *starlark.Thread, fn *starlark.Builtin, a
 	return starlark.None, nil
 }
 
+func (s *tiltfileState) dockerPruneSettings(thread *starlark.Thread, fn *starlark.Builtin, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
```
all new functions should be implemented as extensions, see this pr for an example:
https://github.com/windmilleng/tilt/pull/2330

it should be pretty straightfoward

maiamcc replied:
will do in a separate PR to reduce size of this one

maiamcc commented:
oh whoops i just realized i left the state locked for all of pruning, i was gonna fix that.

nicks commented on internal/engine/dockerprune/docker_pruner.go:
```diff
@@ -35,57 +36,77 @@ func NewDockerPruner(dCli docker.Client) *DockerPruner {
+
+	// Don't prune WHILE we're building something, in case of weird side-effects.
+	// TODO: more robust check here
+	if state.CurrentlyBuilding != "" {
```
i'm pretty sure this won't behave as you expect. this will just fire as soon as the first build completes. i would probably use BuildController's needsBuild function

nicks commented on internal/cli/down.go:
```diff
@@ -48,6 +48,8 @@ func (c *downCmd) down(ctx context.Context, downDeps DownDeps) error {
 		return err
 	}
 
+	tlr.DockerPruneSettings.Disabled = true // don't docker prune on tilt down
```
i wonder if we should do the opposite: DockerPruneSettings.Enabled, which default to false unless the Tiltfile has run and this is a 'tilt up'

maiamcc replied:
are you saying that the default for the USER should be that DockerPrune is disabled unless they explicitly enable it, or just that in the CODE we should default to `disabled` unless it's a `tilt up`?

nicks replied:
this is just a comment about how to organize the code

maiamcc replied:
k, done!

nicks commented:
high-level question: my understanding is that this still has a major bug where if you haven't run Tilt for a few days, and you run 'tilt up', it will delete all your cached images. is that still an issue?

i'm surprised we're not prioritizing that over making it configurable? That bug seems like a launch-blocker (since it makes Tilt unusable), whereas the config issues do not

maiamcc commented:
@nicks think i addressed all your comments except:
- this won't work for cluster users -- will address in a separate PR, and be sure to merge the two at the same time
- implement the tiltfile func as an extension -- will address in a separate PR.

Would love a re-review!

maiamcc commented:
👋 bump!

jazzdan commented on internal/engine/dockerprune/docker_pruner.go:
```diff
@@ -35,57 +38,80 @@ func NewDockerPruner(dCli docker.Client) *DockerPruner {
 	}
 
-	if dp.started {
+	if err := dp.sufficientVersionError(); err != nil {
```
Should we log something in verbose in this case?

maiamcc replied:
will do

nicks commented:
ah, sorry, this PR has gotten way too big to review, i thought the plan was to break it up?

jazzdan commented on internal/store/engine_state.go:
```diff
@@ -85,15 +85,7 @@ type EngineState struct {
-
-func defaultDockerPruneSettings() DockerPruneSettings {
-	return DockerPruneSettings{Disabled: true} // for now, always disabled until we can configure it
+	DockerPruneSettings model.DockerPruneSettings
```
This looks weird to me. What's happening here? Is this returning the empty value of `model.DockerPruneSettings`?

maiamcc replied:
Yep: if we don't bubble the TiltfileLoadResults all the way through the engine state, then we get the empty value of `DockerPruneSettings` (where `Enabled = false` -- so in case like `tilt down`, we're not running the pruner at all).

jazzdan replied:
Can we be explicit about what the default values are rather than returning empty values for everything?

jazzdan commented on internal/tiltfile/tiltfile_state.go:
```diff
@@ -1372,6 +1384,34 @@ func (s *tiltfileState) setTeam(thread *starlark.Thread, fn *starlark.Builtin, a
+	}
+
+	if disable && (intervalHrs != 0 || numBuilds != 0 || maxAgeMins != 0) {
+		return nil, fmt.Errorf("can't disable Docker Prune (`disabled=True`) and pass additional settings")
```
Why not?

maiamcc replied:
> ah, sorry, this PR has gotten way too big to review, i thought the plan was to break it up?

idr us discussing that but sure, can do

maiamcc replied:
cuz it doesn't make sense? if we're not ever running docker prune, why would we allow you to configure settings for it? (i don't feel strongly about this tho.)

jazzdan replied:
I could imagine it being annoying if you wanted to do something like this:

```python
pruner_disabled = read_config()
docker_prune_settings(disable=pruner_disabled, interval_hrs=2)
```

With how the API is now you'd have to instead rwrite
```python
pruner_disabled = read_config()
if pruner_disabled:
  docker_prune_settings(disable=pruner_disabled, interval_hrs=2)
```

Not a big deal, but that's my logic.

maiamcc replied:
mmm fair.

maiamcc commented:
> ah, sorry, this PR has gotten way too big to review, i thought the plan was to break it up?

idr us discussing that but sure, can do

maiamcc commented:
closing this PR in favor of three smaller ones:
- #2340 (move relevant `buildcontrol` funcs to their own package)
- #2344 (use `state.DockerPruneSettings` to determine when/how we prune)
- #2345 (configure `DockerPruneSettings` via the tiltfile)
